### PR TITLE
First pass on using a Tor client 'natively'.

### PR DESCRIPTION
**Pull request by ipeirotis**
Working on issue #94

### CODE REVIEW

File: scholarly/_navigator.py [hunk 1]
```diff
@@ -149,12 +136,44 @@ def _use_proxy(self, http: str, https: str):
         """ Routes scholarly through a proxy (e.g. tor).
             Requires pysocks
             Proxy must be running."""
-        self.logger.info("Enabling proxies: http=%r https=%r", http, https)
+        self.logger.info("Enabling proxies: http=%r https=%r", http, https)        
+        self._http_proxy = http
+        self._https_proxy = https
+
+    def _use_tor(self, tor_cmd=None, tor_sock_port=_SCHOLAR_TOR_SOCK_PORT, tor_control_port=_SCHOLAR_TOR_CONTROL_PORT):
```
Comment (silvavn) [inline scholarly/_navigator.py:143]: Headers with globals not recommended.

File: scholarly/_navigator.py [hunk 2]
```diff
@@ -149,12 +136,44 @@ def _use_proxy(self, http: str, https: str):
...
+        '''
+        Starts a Tor client running in a schoar-specific port, together with a 
+        scholar-specific control port.
+        '''
+        self.logger.info("Starting tor as the proxy")
+        
+        if tor_cmd is None:
```
Comment (silvavn) [inline scholarly/_navigator.py:150]: This not necessarily works. There are several comments on stem about incompatibilities with windows. Also tor .exe can be located elsewhere.

File: scholarly/_navigator.py [hunk 3]
```diff
@@ -27,28 +29,11 @@
...
-_TOR_CONTROL = None
-
-
-if sys.platform.startswith("linux"):
-    _TOR_SOCK = "socks5://127.0.0.1:9050"
-    _TOR_CONTROL = 9051
-elif sys.platform.startswith("win"):
-    _TOR_SOCK = "socks5://127.0.0.1:9150"
-    _TOR_CONTROL = 9151
+_SCHOLAR_TOR_SOCK_PORT = 9600
```
Comment (silvavn) [inline scholarly/_navigator.py:35]: perhaps move this inside the class

**Comment by ipeirotis** [comment]
@silvavn Check the latest version. I am happy with the current structure, and is very transparent to the user.

By default, we do not try to use a proxy. By default, the user can use a direct connection.

If the user decides to use a proxy (either a non-refreshable Tor, or a regular proxy), they can use the `use_proxy` command and set the proxy variable.

If the user has a Tor server that is configured with a control port, they can use the `setup_tor` command to specify the control port for the Tor server and the authentication password.

Finally, if the user wants scholarly to launch a Tor process, then they can use the `launch_tor` function, by specifying where the tor executable is, and which ports to use for the socks proxy and the control port for Tor.

You will notice that I took out any code from the library that guesses Tor ports or locations for the executables. I moved these in the tests, and probably should move them to documentation examples and have them run as tests using doctest.

I assume, we should also expose the `setup_tor` and `launch_tor` functions in the scholarly module, for consistency with the `use_proxy`.

**Comment by silvavn** [comment]
Please, fix and merge develop according to the latest changes applied by this branch.

As I requested above, if we keep merging other branches directly into master, the develop branch will be constantly outdated. Please, let's maintain the developer branch ahead o, or at least the same level, of master. That way we don't have development conflicts. Another good practice would be to always wait for another reviewer to merge into master.

**Comment by ipeirotis** [comment]
Sorry, my mistake. I will target `develop` before merging to master moving forward. Let me see if I can setup proper permissions for the branches, to avoid such issues in the future.

**Comment by silvavn** [comment]
Can we rollback this merge? 
Its not working consistently and I would prefer that we make it stable before the release.

**Comment by ipeirotis** [comment]
The previous version was breaking Mac completely, hence my merge earlier in the morning. 

The `launch_tor` still needs work, but going through scholarly.setup_tor works without issues.

**Comment by ipeirotis** [comment]
Having said that, I am fine reverting to the earlier state for master, and I can incorporate the changes from this PR into `develop` and then to `master`

**Comment by silvavn** [comment]
I have already incorporated your changes into develop and have started
documenting everything. i will have a readthedocs page running in a few
hours :) I am having difficulties making it running on windows as well
thats why I suggested the revert.

Victor Silva

This message was typed and sent from a mobile phone. Please excuse any
typos.

**Pull request merged**